### ISSUE TITLE
Linker cannot remove unused static vars which have an initializer

### ISSUE THREAD

**Issue by vargaz** [mono/linker#368]
Testcase:
```c#
using System;

public class Tests
{
	public static int foo = 5;

	public static void Main () {
		Console.WriteLine ("Hello!");
	}
}
```
Run with:
monolinker -a bug.exe
If = 5 is removed, the 'foo' variable is also removed.

**Comment by marek-safar** [comment]
This limitation comes from linker not rewriting IL code. The code ends up injected in the ctor (or multiple ones) and we'd have to find the right IL segment in the body to rewrite presumably with nops.

**Comment by vargaz** [comment]
I just noticed this pattern in the class libraries, like in type.cs:
        public static readonly MemberFilter FilterAttribute = new MemberFilter(__Filters.Instance.FilterAttribute);
        public static readonly MemberFilter FilterName = new MemberFilter(__Filters.Instance.FilterName);
        public static readonly MemberFilter FilterNameIgnoreCase = new MemberFilter(__Filters.Instance.FilterIgnoreCase);

These 3 are delegates which keep the the methods they reference alive. No idea how much this would save, don't have the tooling to compute it.

**Comment by marek-safar** [comment]
Fixed the easy case

**Comment by mrvoorhe** [comment]
fyi, to wrap up and capture some notes I have on this.

I implemented a prototype in our UnityLinker that strips out the IL to initialize fields that are never loaded, which then let's the field get stripped.  The size reduction is more significant than the beforefieldinit approach, but still relatively small.

For a `Console.WriteLine("Hello world")` program referencing only `mscorlib` the `beforefieldinit` approach yielded around a 0.06% savings.

My prototype (which did not have the `beforefieldinit` change) reduced the size by 0.6%.  The number may vary up or down a bit as I finalize the implementation.  I may be over stripping some cases and I saw some other cases I wasn't detecting but could probably be stripped.

The two approaches together likely yield slightly better results than either alone.  I saw a few fields stripped by the `beforefieldinit` approach that my prototype did not strip.

Both approaches yield larger savings as you pull in more class library assemblies.  Don't recall exact numbers off hand.

Even 0.6% is small so I've put this on the back burner for now. Your other issue https://github.com/mono/linker/issues/367 yields much more considerable gains.  Looks like potential of over 1% and doesn't require stripping IL from method bodies.
